Arcade-powered source-build doc (draft review)

dagood opened the pull request.

dagood committed:
WIP arcade-powered source-build doc

dagood committed:
Feedback; end result; some wording; add chunk

dagood committed:
Clarify point about respin decision factor

dagood committed:
Clarify chunks, add opinion on chunks

dagood committed:
Clarify speculative build difficult and useful

dagood committed:
Clarify source-build func need for intermediates

dagood committed:
Add detail docs: up to speed w/ planning meeting

dagood committed:
rm unused hypothetical chunks

The hypothetical chunks don't actually follow the requirements and would be pointless to show.

dseefeld commented on Documentation/planning/arcade-powered-source-build/README.md:
```diff
@@ -0,0 +1,155 @@
+     an SDK build failure.
+
+   * Breaks in source-build can be fixed immediately by repo owners, just as any
+     other build failure, rather than fixed by source-build maintainers days
```
I think "days after" is optimistic.  It's usually weeks, because we need to wait for a coherent build before even attempting to build.

dagood replied:
Ah, I was originally thinking days since we get a coherent build, but that's not what I'm writing about here really... changing to "weeks to months", the worst case being someone making a change in `master`--since we have to wait for a *very* long time for those to bubble up.

dseefeld commented on Documentation/planning/arcade-powered-source-build/source-build-in-pipeline.md:
```diff
@@ -0,0 +1,41 @@
+The following rough diagram shows a pipeline with official build and validation
+jobs enabled:
+
+> ![](img/rough-pipeline.png)
```
Comments on the pipeline image:
- The "Source-buildable tarballs" arrow indicates when we want to generate a tarball for distro maintainers.  I think it should be clear that this is only an option in core-sdk. 
- I'm looking at the 4 build types on the bottom of the image.  It's unclear whether these are always built when this individual component is built or if they are options.  I think these are the validation jobs.  If so, it would be good to call this out  i.e. required vs. optional.

dagood replied:
> The "Source-buildable tarballs" arrow indicates when we want to generate a tarball for distro maintainers. I think it should be clear that this is only an option in core-sdk.

I was thinking that we shouldn't necessarily limit ourselves to that, however we *would* need to speculatively build an SDK to actually produce a source-buildable tarball that can bootstrap. Probably not useful. I agree, adding core-sdk specificity.

> I'm looking at the 4 build types on the bottom of the image. It's unclear whether these are always built when this individual component is built or if they are options. I think these are the validation jobs. If so, it would be good to call this out i.e. required vs. optional.

👍

dseefeld commented on Documentation/planning/arcade-powered-source-build/in-arcade.md:
```diff
@@ -0,0 +1,121 @@
+special requirements.
+
+At this point in the build, we also have all we need to create a
+source-buildable tarball. We should allow one to be created as early as possible
```
Tarball generation would still require an initial production build through all repos, though, for 2 reasons:
1) the reference-only package ilasm/ildasm process
2) Source generation in coreclr.

Also, I'm not sure where the prebuilt detection logic goes in this scenario.

dagood replied:
I think we should fix these so we don't need a full production build. Reference-only prebuilts should be tracked (so we don't need to discover them via performing a build), source generation in coreclr is trickier but I think we should track it in source-build to formulate ideas. I'll note these in the issue tracking fullbuildless tarball.

> Also, I'm not sure where the prebuilt detection logic goes in this scenario.

Yeah, I neglected this... in terms of ordering it needs to go in Publish to be as sure as possible that all restores for the repo are complete. (Since I don't think all restoring is guaranteed to happen during arcade's Restore phase, unfortunately.)

dseefeld commented on Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md:
```diff
@@ -0,0 +1,36 @@
+  * `dotnet-runtime-5.0.0-alpha.1.12345-linux-x64.tar.gz`
+  * ...
+
+Official builds publish the monster nupkg, and downstream repos restore the
```
I think more is needed here on 2 things.  
1. When are "official" intermediate packages built and pushed to the feed?  I'm assuming this happens on the rolling build for the repo.  Also, when this fails, how is it recovered and how does it impact downstream builds, i.e. if the package doesn't exist, does the downstream repo build use prebuilts or fail on a missing dependency? 

2. Versioning for these?  I see you specified the version for contained nupkgs as the version of the intermediate package.  That should maybe be specified.

dagood replied:
> 1. When are "official" intermediate packages built and pushed to the feed? I'm assuming this happens on the rolling build for the repo.

The official build is a specific type of build--depending on the repo it may be per commit, on a schedule, etc.

> Also, when this fails, how is it recovered and how does it impact downstream builds, i.e. if the package doesn't exist, does the downstream repo build use prebuilts or fail on a missing dependency?

When it fails, the official build fails--this way we never need to handle it being missing. I would say we shouldn't try to handle it being missing because that just means people are free to leave their source-build job broken without caring.

> 2. Versioning for these? I see you specified the version for contained nupkgs as the version of the intermediate package. That should maybe be specified.

Good idea. Yep, it just follows the standard Arcade version for the repository. (It shouldn't ever stabilize, since it's a transport package--this helps some with respinning.)

dseefeld reviewed (approved):
Thanks, Davis.  I read through all the docs and left some comments.

dagood committed:
Review feedback

"Days after" => "weeks to months after"

Limit tarball generation to Core-SDK

Clarify official build vs. validation/rolling build diagrams

Add nuance to early tarball generation: mention roadblocks

Add prebuilt detection

Clarify source-built intermediates must build for official build to pass

Mention versioning of the source-built intermediates nupkg
